### ISSUE TITLE
Question - too much time spent on scheduleDrainBuffers for reading operations

### ISSUE THREAD

**Issue by alejo-17** [ben-manes/caffeine#270]
I'm using Modeshape JCR implementation, and it internally uses caffeine cache configured on the following way:
`Caffeine.newBuilder().maximumSize(workspaceCacheSize).executor(Runnable::run).build();`
I noticed that for each read operation is trying to perform some cleaning up operations, represented on this method: scheduleDrainBuffers(). It internally performs the operation on the current thread. The problem that I see is that when Modeshape calls many times to read the cache, and each scheduleDrainBuffers is taking in one example about 128 ms, which seems to be high. Why is this operation required when reading the cache?
I attached some images showing the time spent on executing clean up tasks:
<img width="1013" alt="executionpath" src="https://user-images.githubusercontent.com/43454632/45851844-82645d80-bd02-11e8-9801-efa38d427875.png">
<img width="1138" alt="executiontime" src="https://user-images.githubusercontent.com/43454632/45851851-898b6b80-bd02-11e8-862d-a3162a9b4182.png">

I appreciate your help.

Thanks

**Comment by ben-manes** [comment]
I agree 128ms is very high, as it usually executes in terms of nanoseconds. Sometimes a sampling profiler will penalize fast but frequently called methods incorrectly, so it may not be the problem.

The design reason is because we want to maintain eviction policies that are cheap but not concurrent. For example LRU is merely moving the current node and appending it to the end of a doubly-linked list. That's a cheap O(1), except it cannot be done efficiently lock-free or with naive locking. Doing so creates a lot of contention, which is more expensive than the operation itself. Instead we record into a buffer and replay the operations when its full. That way we can batch the work under the lock and acquire it using a `tryLock` since the buffer holds their work if unsuccessful. This approach (similar to a database's write-ahead log) allows for implementing an LRU-type cache with a small overhead.

I'd probably need more information to help you debug this performance issue. I would recommend switching to a `computeIfAbsent` style call instead of a `get/put` to avoid the thundering herds problem. If you can reproduce it in a smaller benchmark that might help, as my own stress tests never showed this. If you look through the code, you should see it shouldn't be doing much expensive work in that configuration.

**Comment by ben-manes** [comment]
I would also recommend testing with the current version, `2.6.2`. It appears they are use `2.3.3`. It should be api compatible.

**Comment by ben-manes** [comment]
You might be affected by https://github.com/JCTools/JCTools/issues/135, which was fixed in `2.3.4`. That race condition caused the consuming thread to be stuck in the write queue and may explain why you're seeing the drain not function properly.

**Comment by alejo-17** [comment]
Thanks for the useful information and the prompt answer!

I will try upgrading the caffeine version to `2.6.2` and see if that fixes the issue. If not, I will get more evidences to troubleshoot this issue.

I appreciate your help!

**Comment by alejo-17** [comment]
I have tried by upgrading caffeine version to `2.6.2` and also using `computeIfAbsent` style, but it looks like these options didn't fix the issue. Is it possible that if the cache size is not enough, the `scheduleDrainBuffers()` takes more time than the expected due to the eviction process? Should I consider adding Time-based eviction mechanism? How should that work if `Runnable::run` is used as an executor?
For example, in the first 15 minutes of a test with concurrency I see the slowness of `scheduleDrainBuffers`  (about 15 seconds for one request):

<img width="1021" alt="15min-exectime" src="https://user-images.githubusercontent.com/43454632/46045597-be713700-c0e3-11e8-9ec4-e15bde07218a.png">


<img width="880" alt="onerequest-15min" src="https://user-images.githubusercontent.com/43454632/46045500-810ca980-c0e3-11e8-9ec7-a249435037f3.png">

Please let me know your comments/thoughts.

Thanks.

**Comment by ben-manes** [comment]
hmm, that seems really weird. 

I know early versions of JDK8 had a broken ForkJoinPool, which could cause trouble (but you're using the same thread). That and the JCTools bug caused the write queue to not be drained, or be a lot more expensive with some back-pressure fixes on our end. Otherwise at your configuration and release, it should be very cheap. If you look at both methods, they aren't doing much. Do you see everything return to a steady state or is it stuck, e.g. in an infinite loop or blocked on a lock? It's weird because its more like the maintenance isn't being run (e.g. due to the earlier bugs) and things are stalling. That would make sense if the JCTools bug was still present, e.g. if you can verify it's really hitting 2.6.2.

Time-based operations don't use a scheduling thread and expired items are evicted during the maintenance process. But I don't know if that could help here.

**Comment by alejo-17** [comment]
I have validated the jar used was `2.6.2`. Actually, I realized that the first test results I had sent were performed with caffeine version `2.3.5`, which contains the JCTools fix. Is there any difference between these two versions?
By the way, it seems like everything returns to a steady state, since the service is only taking more time to respond, but I can corroborate that with thread dumps. Do you think it's expensive to use the same thread for maintenance tasks?

**Comment by ben-manes** [comment]
Do you think it would be possible to isolate this to a stress test that I can run? I’m not sure how to help investigate.

Usually the same thread executor is fine. Mostly the executor is useful for callbacks like the removal listener, where the same thread could increase tail latencies.

Is your workload write heavy? The only time the drain should be doing a lot of work in your case is if it can’t catch up with the writers.

**Comment by alejo-17** [comment]
Yes, my workload writes heavy. For example, for one request the cache is hit on this way: 
- putIfAbsent -> 193k times approximately 
- get -> 529k times approximately

In addition, I have been testing without using the same thread (`ForkJoinPool`) for the executor and I noticed that the `performCleanUpTask` is being executed less times than using `Runnable:run`. Is there a difference on calling this task by using a different executor? For example, for a different test, but in the same request I have the following:

- Same thread: 140k times approximately
- ForkJoinPool: 43k times approximately

I will have a sample stress test once I could replicate this out from Modeshape.

Please let me know your comments/thoughts.

Thanks.

**Comment by ben-manes** [comment]
Thanks a lot for your persistence on this issue. If it's helpful, I have a simple [Stresser](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/test/java/com/github/benmanes/caffeine/cache/Stresser.java) that I use to detect deadlocks and such things.

The main difference should be delays due to context switching. For same-thread it can execute the task directly when scheduling the work. For the thread pool, a thread must become available and pick up the work. Since there is an additional delay, such as if you use FJP for application tasks, this would give a larger opportunity for writers to append to the buffer and the FJP thread to perform a larger batch.

Assuming there is no data structure bug, I can think of two ideas we might explore:
1. There might be a high level of contention on the hashmap causing evictions to be slow. `ConcurrentHashMap` uses bin locks, which are coarser than per-entry. This penalty would be on your calling thread, though ideally should be fast if no expensive computation. You might try increasing the initialCapacity, as the larger the map the more locks there are.
2. If the write-rate exceeds the drain rate, then the runner is penalized by draining all up to `WRITE_BUFFER_MAX` (~ 128 x npus). For the same-thread, you might argue it should be better amortized across callers. However, at such a thrashing `afterWrite` should degrade to blocking on the lock and thus not contending with the draining thread. This did seem to increase throughput and provide a good failsafe backoff. But we could try decreasing this parameter to see how it effects things.

If you have any questions on the design or the code, please let me know. I wrote an [overview](https://github.com/ben-manes/caffeine/wiki/Design) since I know its a lot of code to wade through.

**Comment by ben-manes** [comment]
Can you enable stats and check the hit rate?

**Comment by Siva-R** [comment]
Ben-Manes, the picture in the first comment which Alejandro had was related to a read operation, there were no write operations in his test, but still the execute method is being called several times. Is it not an expectation that, when the execute method is called for the first time, it will free up enough nodes to avoid calling it several times?

**Comment by ben-manes** [comment]
That's a good point, @Siva-R. The execute is only called within a `ReentrantLock`, so all other threads would skip over the call since the work is being handled. This would only make sense if there are multiple cache instances.

https://github.com/ben-manes/caffeine/blob/f107fbc8b93883ec5513916c6c7fb5a819e6b498/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L184

https://github.com/ben-manes/caffeine/blob/f107fbc8b93883ec5513916c6c7fb5a819e6b498/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1065-L1088

The only other place where `execute(Runnable)` is called is by a removal listener, which is often outside of the lock, but supposedly that's not being used.

https://github.com/ben-manes/caffeine/blob/f107fbc8b93883ec5513916c6c7fb5a819e6b498/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L281-L296

**Comment by alejo-17** [comment]
Thanks for suggesting those ideas. I have a couple of questions about them:

- About `initialCapacity`, are you suggesting I should set that value with a similar value of maxSize? For example if the cache size was 50k, would I set that value to 50k or maybe 25k?

- Regarding `WRITE_BUFFER_MAX`, I don't understand how that is related to the drain buffer, could you please provide an example of what you're proposing?

By the way, I'm working on enabling stats inside Modeshape code, so as soon as I get them, I will provide you with that information.

Thanks

**Comment by ben-manes** [comment]
1. Yes, it would be interesting to see if sizing it upfront changes things. `ConcurrentHashMap` should grow the table as the cache is populated, but will start out small with few locks. It's purely speculation because your profiler runs should have highlighted that as the cause.
2. There are two buffers: read and write. The read buffer is lossy as the information is just a hint to the policy of what to keep. The write buffer isn't, e.g. adds trigger evictions, so this grows to capacity and then blocks. The drain of the write buffer is capped by an internal constant `WRITE_BUFFER_MAX`, so that if writes keep coming in the caller isn't stuck forever (e.g. if same-thread and delaying user-facing responses). So a random guess is if its too lenient and a smaller cap would better amortize this across callers. The same amount of work would be performed, just spread out more granularly. (This would require a custom build to experiment with)

https://github.com/ben-manes/caffeine/blob/f107fbc8b93883ec5513916c6c7fb5a819e6b498/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L145-L146

https://github.com/ben-manes/caffeine/blob/f107fbc8b93883ec5513916c6c7fb5a819e6b498/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L1251-L1264

**Comment by ben-manes** [comment]
yeah, I really don't recall the reasons. I suspect it's because a lot of various map methods might be considered for recording a hit or miss, but its not clear if desirable. For example `put` or `replace`, as well as the iterators. I wouldn't feel comfortable changing this much without @yrfselrahc and/or the Guava team's opinion to do so, too.

You could use Guava's `ForwardingMap` as a quick workaround to delegate to `getIfPresent` or record stats manually.

My thinking in this regard perhaps your workload underperforms with TinyLFU, e.g. by being heavily recency skewed. If the cache is always missing then writes are far more common, which might be the real problem here. If so, then we can discuss the recent research to fix that (#266) and the short-term workaround (#263).

**Comment by alejo-17** [comment]
Thanks for the information. I'm working on getting statistics manually.

By the way, I have one concern regarding the optimum cache size that I use. I got a count from my database and I saw about 2 million of rows that can be cached, since Modeshape caches all by default, and I have 50k as a max cache size. I tried by increasing it to 500k, but I started to see lots of suspension times, so the performance has decreased. Is caffeine able to work with larger size cache? Does it require to have more JVM memory? I have currently 2.5 GB as max configured.

**Comment by ben-manes** [comment]
It may be a memory limit. Java is difficult to size for, since the byte size is JVM and platform specific. To some extent you can make guesses using the `Weigher` to assist, e.g. [JAMM](https://github.com/jbellis/jamm) for a rough (but inaccurate) guess. [JavaObjectLayout](http://openjdk.java.net/projects/code-tools/jol/) is the closest match, but isn't really meant for anything beyond experiments. Caffeine has its own metadata so that also adds some overhead per entry, too. When it gets too large then it might be worth exploring multiple layers, etc. instead.

The backing hash table is `ConcurrentHashMap` which is designed for a large number of entries. However that's more about how it degrades big-O than byte sizes. An off-heap or alternative might become necessary, which gets out of scope for this project.

**Comment by alejo-17** [comment]
I was able to get some statistics after running a single test for one features and I got the following numbers:
- Get call: 243,244 times.
- PutIfAbsent call: 4752 times.

Therefore, it says that the hit rate is around 96% and 98%, which concludes that the cache is not always missing the entries. Also, in a performance test the % of writing operations is less than reading operations:

<img width="1009" alt="hotspots-afterread-afterwrite" src="https://user-images.githubusercontent.com/43454632/46320029-05f33980-c5a2-11e8-9099-1e6bb943f1ba.png">

So reading operations are taking time to drain buffers.

Please let me know your comments/thoughts.

Thanks.

**Comment by ben-manes** [comment]
Thanks! In that case then I'm not sure. Can you open up the `afterRead` block to see the breakdown?

The concurrent calls should be okay, because they accumulate into a striped ring buffer. That should be the limiting factor where contention can occur, as its shared state that we CAS on. The striping should help alleviate that, so its usually not visible in a profile given application overheads. Then `scheduleDrainBuffers` should be cheap because its a `tryLock` that is allowed to fail and skipped over. It sounded like the `maintenance` was a bottleneck, but if you read the code it should be simple code under the lock that reads from the buffer and does a few linked list operations. So all of these should be very cheap, at least compared to the alternative of reads going through an exclusive lock.

**Comment by alejo-17** [comment]
I'm attaching the caller breakdown of `scheduleDrainBuffers` operations (from a different performance test):

<img width="1205" alt="scheduledrainbuffers" src="https://user-images.githubusercontent.com/43454632/46321389-9fbde500-c5a8-11e8-99bc-0f672b21983d.png">

The class is referencing the lambda of the executor: `org.modeshape.jcr.cache.RepositoryCache$$dtt$$Lambda$283.534425031`

Also I have found a different hotspot, with a short total execution time, but I'm not sure if that is related to the issue:

<img width="1205" alt="invoke_mt" src="https://user-images.githubusercontent.com/43454632/46321443-f0354280-c5a8-11e8-936f-cf656cf39b80.png">

The problem is that Dynatrace is not showing in which part of the execute is the bottleneck, so I don't have that precision unfourtanetly.

**Comment by ben-manes** [comment]
`expirationTicker` is a no-op, so the second trace looks like its incorrectly assigning the blame. In your case of no expiration it returns `DisabledTicker` which is always `0L`, otherwise its `System.nanoTime()`. To be fair that can be a bottleneck on non-Linux machines, but again application code usually dominates.

Otherwise I really don't know. The code for `afterRead` is not too expensive. The `readBuffer` will have contention that is optimized for, and `scheduleDrainBuffers` uses a `tryLock` so penalizes only 1 thread.
https://github.com/ben-manes/caffeine/blob/f107fbc8b93883ec5513916c6c7fb5a819e6b498/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L831-L848

**Comment by alejo-17** [comment]
Thanks for the information! I will get thread dumps to see if there are blocked threads. Just to be curious, what are the conditions to skip the scheduleDrainBuffers call?

https://github.com/ben-manes/caffeine/blob/f107fbc8b93883ec5513916c6c7fb5a819e6b498/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L844-L846

**Comment by ben-manes** [comment]
It’s skipped if (a) a drain is taking place, (b) a drain is not required (write buffer is empty) and the read buffer isn’t full. Due to a state machine this is a simple switch on an enum rather than poking at things, so cheap to figure out.

**Comment by alejo-17** [comment]
Thanks for the information. What I also noticed is that I have two types of `execute` method in my reports for only one request:

One represented by this class: `org.modeshape.jcr.cache.RepositoryCache$$dtt$$Lambda$283`, and in summary they are executed fast (89 ms in total):
<img width="1155" alt="purepath-executefast" src="https://user-images.githubusercontent.com/43454632/46361571-a0e82400-c633-11e8-9c69-53c62ec5181a.png">
<img width="1191" alt="methodhotspot-executefast" src="https://user-images.githubusercontent.com/43454632/46361575-a6456e80-c633-11e8-867a-2546830de3f4.png">

The other one represented by this class: `org.modeshape.jcr.cache.RepositoryCache$$dtt$$Lambda$283.534425031`, which has the same number as the above ($283), but with a extension number 534425031, so this is taking 3.16 seconds, which is very slow:
<img width="1210" alt="methodhotspot-executeslow" src="https://user-images.githubusercontent.com/43454632/46361577-a8a7c880-c633-11e8-987b-6f0214432a4a.png">
<img width="963" alt="purepath-executeslow" src="https://user-images.githubusercontent.com/43454632/46361583-aa718c00-c633-11e8-8c91-462eaf8452c4.png">

The one which concerns me is the last one, but I'm not able to understand what is the difference between these two lambdas if they are supposed to be doing the same.

Please let me know your comments/thoughts.

Thanks.

**Comment by ben-manes** [comment]
That's odd. You could replace the anonymous class with a nested one, just to give it a name. At least then its clear which lambda its referring to. Obviously with 280+ its too much to do everywhere, but you could do spot checks on the ones you are most concerned about. There might be tools in jitwatch or similar to indicate which lambdas refer to which code block, but I can't think of anything offhand.

**Comment by ben-manes** [comment]
I see `putFast` but that method doesn't exist anymore. There was a fast-path and slow path, which was merged into one. So you definitely are not running 2.6.2, just in case you thought so.

**Comment by ben-manes** [comment]
The only reason `scheduleAfterWrite` could be slow, if directly the contributor, is if the `PROCESSING_TO_IDLE` => `PROCESSING_TO_REQUIRED` failed and had to be retried. That's a CAS, so pretty much either one or two iterations. It shouldn't be a hotspot, if that's the claim of the profiler.

https://github.com/ben-manes/caffeine/blob/fe68b00ac52b303dd20231123d61e7c3e03782ac/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L927-L954

**Comment by alejo-17** [comment]
Yes, sorry about that. It was tested with caffeine `2.3.5` since I didn't see any improvement with `2.6.2`. This is the test I did with last caffeine version:

This is the total execution of `execute(Runnable)` for the whole performance test:
<img width="1210" alt="methodhotspots-2 6 2" src="https://user-images.githubusercontent.com/43454632/46370044-35f61780-c64a-11e8-89df-c1449319f663.png">

This is only for one request of `scheduleAfterWrite`:
<img width="974" alt="scheduleafterwrite-2 6 2" src="https://user-images.githubusercontent.com/43454632/46370117-663db600-c64a-11e8-993b-60135d4364ce.png">

**Comment by ben-manes** [comment]
One experiment you might do is run using Guava, since the APIs are identical in your case. Caffeine should be faster, but if not then that confirms your results. Might help figure this out?

**Comment by alejo-17** [comment]
Yes, I think I have some pending items to test. 

By the way, I have been running the Stresser test locally, but with some changes:
- maxCacheSize = 50,000
- initialCapacity = 50,000
- executor = Runnable::run

**For reading operations:**

---------- 00:01:04 ----------
Pending reads: 29; writes: 0
Drain status = Idle (0)
Evictions = 998,576
Size = 50,000 (max: 1,048,576)
Cache Size = 50000
Lock = [Locked by thread pool-2-thread-16]
Pending tasks = 0
Max Memory = 3,817,865,216 bytes
Free Memory = 219,395,224 bytes
Allocated Memory = 324,534,272 bytes
---------- 00:01:09 ----------
Pending reads: 16; writes: 0
Drain status = Idle (0)
Evictions = 998,576
Size = 50,000 (max: 1,048,576)
Cache Size = 50000
Lock = [Locked by thread pool-2-thread-8]
Pending tasks = 0
Max Memory = 3,817,865,216 bytes
Free Memory = 219,395,224 bytes
Allocated Memory = 324,534,272 bytes

**For writing operations:**
---------- 00:02:54 ----------
Pending reads: 16; writes: 0
Drain status = Idle (0)
Evictions = 568,545,387
Size = 50,014 (max: 4,096)
Cache Size = 50080
Lock = [Locked by thread pool-2-thread-15]
Pending tasks = 0
Max Memory = 3,817,865,216 bytes
Free Memory = 321,357,288 bytes
Allocated Memory = 1,146,093,568 bytes
---------- 00:02:59 ----------
Pending reads: 48; writes: 6
Drain status = Required (1)
Evictions = 586,133,639
Size = 51,518 (max: 4,096)
Cache Size = 51299
Lock = [Locked by thread pool-2-thread-4]
Pending tasks = 0
Max Memory = 3,817,865,216 bytes
Free Memory = 936,253,216 bytes
Allocated Memory = 1,538,260,992 bytes

I'm not pretty sure how to interpret the results, but I can see locks for the executions, can you please help me to interpret these results?

Please let me know your comments/thoughts.

Thanks.

**Comment by ben-manes** [comment]
> can you please help me to interpret these results?

Sure. This shows some of the internal states to see if there is a runaway conditions. So it helps answer a few questions like,
 - Is the cache able to abide by the high watermark (exceed it by a small amount, temporarily) or does it have a runaway leak?
- Is the lock held by a single thread, e.g. is it stuck?
- Are the read or write buffers full and never seeming to be drained?
- Is the state machine (drain status) transitioning or get stuck?
- Are there any crashes due to internal corruption when stressed?

Some questions are answerable in unit tests, such as the small stress via the `MultiThreadedTest`, but a longer case can uncover more issues. It's also scaffolding to quickly hack and explore with based on a bug report, e.g. when I didn't think of a test scenario and it broke. It doesn't really show throughput numbers, though you could tweak it to retain a count of operations performed. All it can really do is indicate if things seem to be running smoothly without hiccups, which appears to be the case.

Pending reads / writes - ring buffer sizes
Drain status - state machine to optimize the buffer draining
Evictions - number of items evicted
Size - current & max size (high watermark)
Cache Size - your addition? the configuration I guess
Lock - which thread holds the evictionLock if draining
Pending tasks - write operations to replay on the policy; an update w/o weight change can be mimicked as a read, which is faster.
Memory - JVM memory use (if stable or runaway)

For determining the average latency (throughput) you can see the JMH benchmarks. That shows the upper bar, which we want better than the application can induce. However, beyond that we know from  Amdahl's law that further speedups won't be helpful. In Caffeine's case, we trade-off some surplus read throughput for trying to reduce the maximum latency (especially on writes), e.g. the long tail response times, to provide more predictable performance. The design decisions also have some benefits for how we implemented certain features, which was an added bonus.

**Comment by Siva-R** [comment]
Ben-Manes,
Couple of questions.
1) Is it not that when ring buffer is full, scheduleDrainBuffers() will be called? 
2) If yes, is there a possibility that because of small ring buffer size, scheduleDrainBuffers() is called frequently? 
3) Will there be an improvement if we increase the ring buffer size? Can you suggest an ideal size depending on the load?
4) Is the ring buffer size dependent on the type of thread pool executor?

**Comment by ben-manes** [comment]
Hi @Siva-R 😃 

> Is it not that when ring buffer is full, scheduleDrainBuffers() will be called?

There are two types of ring buffers.
 - The read buffer captures cache hits and consists of many small ring buffers (e.g. 1 per core) to reduce contention. When full, new events are dropped (lossy) because the data doesn't have a negative impact of the eviction policy (hot items are more frequent, so hits accumulate). This triggers `scheduleDrainBuffers()` when one of the ring buffers is filled.
- The write buffer is a ring buffer used to avoid blocking all threads on the global policy lock. This is not lossy and blocks writers if full. When populated a drain is applied immediately, since this might require an eviction.

> If yes, is there a possibility that because of small ring buffer size, scheduleDrainBuffers() is called frequently?

Yes, the the method will be called frequently but only executed by a single thread at any given time. The work involved should be very cheap and O(1), e.g. an LRU is merely pointer manipulation (access: remove node, move to tail; evict: remove head). The observation is that the work involved is cheap, but blocking on a lock is not (e.g. if you benchmark 1 vs 10 threads access a synchronized `LinkedHashMap` in LRU mode). We use intelligent scheduling to avoid the lock penalty.

> Will there be an improvement if we increase the ring buffer size? Can you suggest an ideal size depending on the load?

A smaller read buffer actually has better performance. This is because adding to the ring buffer is a CAS to acquire the array slot. Since the additions are lossy, when full no work is performed so its free. That's under the belief that `scheduleDrainBuffers()` is inexpensive, but if not then you're right that increasing the buffer size would reduce the frequency of it being executed.

> Is the ring buffer size dependent on the type of thread pool executor?

Nope, it is dependent on the number of cores available to the JVM and amount of contention observed.
 - The read buffer uses 16 element ring buffers and adds additional ring buffers when CASes fail due to contention. This grows up to `4 * ceilingNextPowerOfTwo(Runtime.getRuntime().availableProcessors())`. If you are on a high cpu machine, e.g. 32 cores, then it starts at `0` if unused to avoid wasting memory but grows up to 128 buffers if all are accessing it concurrently.
- The write buffer is a single ring buffer that grows in length, starting at `4` and ending at `128 * ceilingPowerOfTwo(Runtime.getRuntime().availableProcessors())`. In the 32 core case, that means 4096 write operations might be enqueued before blocking. This maximum would only occur on a bulk insert (e.g. cache warming) or many writers outpacing the immediate drain. In the worst case all these writes might be adding new items, so we exceed the maximum size by 4k temporarily. But that should be okay since a 32-core machine would not be memory constrained to that degree, which lets us support a high write throughput by the buffer batching the work under the lock.

I hope that helps. Happy to answer more questions! 😄

**Comment by alejo-17** [comment]
After testing by using the `ForkJoinPool` as an executor instead of `Runnable::run`, I'm no longer seeing long execution times on the execute method, so I'm not sure why the issue happens by using the current thread to drain. However, I still see the `com.github.benmanes.caffeine.cache.BoundedLocalCache.dt_0_expirationTicker_501()` as a hotspot.

<img width="1184" alt="methodhotspot" src="https://user-images.githubusercontent.com/43454632/46547087-7e067b80-c890-11e8-9629-47414e5cfe5c.png">

<img width="1183" alt="purepath" src="https://user-images.githubusercontent.com/43454632/46547102-8e1e5b00-c890-11e8-9a15-3d3f3273780b.png">

The host where the tests are executed has only 4 cores, so I'm not sure if that number is not enough for the buffer's calculation.

Please let me know your comments/thoughts.

Thanks.

**Comment by ben-manes** [comment]
I would recommend putting your own timings to verify the accounting. If you are not using expiration, then the `expiration` ticker returns a singleton (enum) instance via `Ticker.disabledTicker()`. That has a single method that returns a constant. Ideally if inlining and dead code elimination occurs this call is removed entirely. In the case of expiration being enabled, then the ticker field should be inlined for the method, and may inline the implementation as well. If the implementation is slow (as `System.nanoTime()` can be), you can override it with a custom version that is cheaper but less accurate.

It may be that the instrumentation disallows optimizations (e.g. C1 instead of C2) or is assigning blame incorrectly. For example, this may be a GC safepoint. It doesn't make sense from a code perspective and seems to be a runtime aspect.

**Comment by ben-manes** [comment]
@alejo-17 any updates here?

**Comment by CodingFabian** [comment]
![image](https://user-images.githubusercontent.com/176132/56988430-50688300-6b90-11e9-936e-f714c3900f7a.png)

Hi @ben-manes 
I am recycling this issue, as we are observing the same, or a similar issue.
We are on 2.6.2, and this cache gets heavy writes.
The data was recorded over 20 minutes with JProfiler 10.1.5 default sampling settings.
it uses
`executor(Executors.newSingleThreadScheduledExecutor()) `
Thread profiling of the maintenance thread shows its busy about 25% of the time.
I am aware of the sampling/safepoint bias this profiling has, but it seems to me that  
`        executor().execute(drainBuffersTask); ` 
is just invoked way too often.
The surrounding locks and CAS operations look fine upon closer inspection to me.

Since we use
`expireAfterAccess(30, TimeUnit.SECONDS)`
I was wondering if maybe the maintenance task just completes so quickly that all of the writes will continue to reschedule the maintenance over and over again?

We intentionally do not use Runnable:run executor, because when the maintenance runs, it would block too long.

**Comment by ben-manes** [comment]
Hi @CodingFabian,

This is hard to diagnose a fix for, so a few background questions.

The maintenance will only run when a write occurs, a threshold of reads is reached, or an expired entry is found by read-only query (e.g. `getIfPresent`, `computeIfPresent`). We don't schedule any background thread based on a timer. Much of the maintenance work is very cheap, so it rarely should be a bottleneck to run directly. Of course, since it has to atomically remove an entry the `compute` can block due to `ConcurrentHashMap` bin lock, which can be compensated for by increasing the `initialCapacity`. Does setting a higher value help if this is the case?

What do you mean by it would block too long by using `Runnable::run`? This will of course mean other asynchronous actions will be invoked on the calling thread, which could be special cased by delegating them to an executor (e.g. your `RemovalListener` schedules the work).

It may be that a task is rescheduled because there isn't enough time to allow writes to buffer and be performed in bulk. Unfortunately there is not much we can do there, since delaying that would cause its own problems by not evicting promptly. If you don't use any other async work (refresh, listener, etc) then you could disable the executor (`task -> {}`) and invoke `Cache.cleanUp()` on a scheduled thread, or something similar but less hacky like queueing up the tasks and run them on a fixed delay). There will only ever be one in-flight maintenance task, as a new one won't be submitted if running or scheduled.

Otherwise, I'm not sure what to change internally. Fundamentally we should process writes which could require an eviction. We already optimize writes that have no direct effect by scheduling them as reads if possible. We buffer pending writes if we can, rather than blocking on the lock.

I agree a short expiration interval could lead to a lot of evictions, but that's the desired behavior.

My gut feeling is that you supplying a custom executor that batches the tasks to execute at a desired interval would be the best solution. If that works for you, we can add it to the FAQ and I'll try to retain that compatibility with that expectation.

**Comment by CodingFabian** [comment]
Sorry, I forgot to supply that information:
We have configured an eviction listener. If lots of entries are evicted (the way its used is quite bursty, if entries expire they tend to expire at the same time), our removal listener blocks.

We will experiment with batching up the removal listener activities or with batching the maintenance tasks. both sound like reasonable approaches.

I am not claiming that the implementation is incorrect. And I also acknowledge that there will be always edge case :-) Just wanted to supply one datapoint you might want to use, and since a similar issue existed I just reused this one :)
